### PR TITLE
Add clippy lints

### PR DESCRIPTION
**Pull request by TimonPost**
### Description of Changes

Copied all the up-to-date lints we use at embark and put them into the workspace toml such that they apply for every project. Also added lots of documentation to public types because I added a lint for this too. 

- Defines clippylints in workspace root
- Fixes clippy suggestions
- Makes documentation required for public types inside puffin.
- Makes parkinglot always a dependency (can revert if this is not desired)

### CODE REVIEW

File: puffin/src/lib.rs [hunk 1]
```diff
@@ -405,8 +339,9 @@ impl ThreadProfiler {
         new_id
     }
 
+    /// Register a named scope.
     #[must_use]
-    pub fn register_scope(
+    pub fn register_named_scope(
```
Comment (emilk) [inline puffin/src/lib.rs:344]: This is a breaking change that slipped in – was that intentional?

**Review by emilk — APPROVED** [review]
I love it!

Are we sure this doesn't require updating the MSRV of the crate itself?

File: puffin/src/lib.rs (hunk 1 above)
Comment (TimonPost) [reply puffin/src/lib.rs:344]: this function should be hidden from API as it should not be called by users. I will revert it in this PR but follow up with general encapsulation PR

**Comment by TimonPost** [comment]
Oh was not suposed to merge automatically. I'll have a look at your question. Why do you think it requires updating the MSRV of the crate itself?

**Comment by emilk** [comment]
Yeah, annoying bot.

> Why do you think it requires updating the MSRV of the crate itself?

It shouldn't - I just now realized the new sections (`[workspace.lints.rust]`) are in the workspace toml, so it should not at all affect the crates.

**Pull request merged**